# Introduction to Python
*Credits to the Expedia EWE Singapore team (CodeSingapore@expedia.com). Multiple lessons were copied and rearranged from their github repo..*

* **Class:** 
  * Please introduce yourselves
  * State if you have learnt programming before (and if so which language)
  * Tell what you hope to achieve at the end of the course

## Course Overview

* **Programming Language:** Python
* **Data:** We will use the titanic dataset
* **Every week:** You will learn something new in Python to help you answer questions on the titanic dataset
* **At the end of the course:** You should be able to do simple data analysis using Python 

## Why is Python named Python?

[Guido van Rossum](https://en.wikipedia.org/wiki/Guido_van_Rossum) is the creator of Python and also Python's ["Benevolent Dictator For Life" (BDFL)](https://en.wikipedia.org/wiki/Benevolent_dictator_for_life) - recently stepped down as of Jul 2018. When he started working on Python, he was also reading the published scripts from ["Monty Python's Flying Circus"](https://en.wikipedia.org/wiki/Monty_Python%27s_Flying_Circus), a BBC comedy series from the 1970s. Van Rossum thought he needed a name that was short, unique, and slightly mysterious, so he decided to call the language Python.
Python was first introduced in 1990.

## But why are *we* learning Python?

Python is a powerful high-level programming language, yet readable and easy to understand. No matter which [programming language ranking](https://techbeacon.com/programming-language-rankings-which-ones-matter) you refer to, Python is one of the most popular programming languages now. 

To find out more, check out [this article](https://medium.freecodecamp.org/what-programming-language-should-i-learn-first-%CA%87d%C4%B1%C9%B9%C9%94s%C9%90%CA%8C%C9%90%C9%BE-%C9%B9%C7%9D%CA%8Dsu%C9%90-19a33b0a467d) which includes a cool diagram on which programming language to learn first.

## Let's create your library to load your lessons!

* Visit https://github.com/kelvnt/python-intro
* At the bottom left of the screen, click on **Launch AzureNB badge** -> **Clone** -> Sign-in using your own email -> **Clone** again
* Click on **Setting** -> Deselect **Public library** -> **Save** -> Make sure that the **Private** badge is displayed below library's name
* Click on **Run** -> Change browser URL path from **/nb/tree** to **/nb/lab**
* Bookmark this URL

## Wait.. I don't need to install anything? 

No, you don't. We are hosting everything on **JupyterLab** (a web-based user interface that allows you to run your code in a flexible and integrated manner) and **Azure Notebook** (which provides free online access to Jupyter notebooks and JupyterLab).

If you would like to install an [Integrated Development Environment (IDE)](https://en.wikipedia.org/wiki/Integrated_development_environment) for your own Python exploration, I'd recommend using Spyder which comes with the Anaconda installation that you can consider for your own laptop.

## Done? Let's talk about different File Types

* A file (like this one!) which has a mixture of text as well as Python code is an iPython notebook, and has the file extension *.ipynb*
* A file with only Python code is a Python script, and has the file extension *.py*

## We will teach you 3 methods to run Python code
1. Run python code within an iPython notebook (good for code snippets)
2. Run python command from the *Python* terminal (similar to method 1, just not so pretty)
3. Run python script from the terminal (for running entire programs)

## Method 1: Run Python code within iPython notebook

In [1]:
# Hello, this is a code cell.
# Comments start with # symbol, and don't get executed. 

# Select this cell and click on the play button above to run

print("Hello World")

Hello World


## Awesome, you've just written your first Python code!

Want to try some more? Run the following cells!

In [8]:
1 + 1

2

In [9]:
7 - 4

3

In [10]:
2 * 3

6

In [11]:
5 / 2

2.5

## Now, it's your turn to add a new cell!

* Click on "+" button above to add a new cell
* Add a new Code cell and create your own simple print command or math operations in Python

In [12]:
print("Kelvin")

Kelvin


In [13]:
print("k")

k


## Method 2: Run python command from the *Python* terminal 

* Click the "+" button on the navigation pane (left column), and a new Launcher tab will appear
* Select Terminal and run the following commands

> nbuser@nbserver:~\$ python

* This will launch Python terminal where the prompt looks like ">>>"
* Try the same commands as you tried previously, and you'll find the same results!

> print("Hello world")

> 1+1

## Method 3: Run Python script from the terminal

### Let's create the Python script first!

* Click the "+" button on the navigation pane (left column), and a new Launcher tab will appear
* Select TextEditor, and a new file *untitled.txt* will be created
* From the navigation pane, right click and rename the file to *hello.py* (replace .txt too). Check if the file icon has changed to the Python icon
* Add the following content in *hello.py*: print("Hello, world!") 
* Save from menu option *"File > Save Python File"* or press Ctrl+S (Windows) / Cmd+S (OSX)

### Now to run the Python script!

* Click the "+" button on the navigation pane (left column), a new Launcher tab will appear
* Select Terminal and run the following commands

> nbuser@nbserver:~\$ cd library
>
> nbuser@nbserver:~\$ python hello.py

* You should get the same results as above!

## Variables

### What is a variable?

Variables are used to store information to be referenced and manipulated in a computer program. In simple words, they are containers that hold information.

Their sole purpose is to label and store data in memory. This data can then be used throughout your program.

### Why are they important?

They provide a way of labeling data with a descriptive name, so our programs can be understood more clearly by the reader and ourselves.

Also, by storing data in a variable, you can access this data anywhere in your program without duplicating it. Therefore, it's easier to read and manipulate the data throughout your program

### How to assign a value to a variable

```python
myVar = 42
```

That's it! Easy right?

### Inputs / Outputs

Many times in this lesson you will read words like "inputs" or "output". Just a quick clarification on what these two terms refer to:

- Inputs are all the data/variables that the program stores or receives

```python
# My inputs:
aVar = 4
anotherVar = 4 * 8
```
Inputs can also come from the user as we'll learn further down in this lesson.

- Outputs are everything that a program returns or print out.

For instance:

In [14]:
# My inputs:
aVar = 4
anotherVar = 4 * 8

# My outputs:
print(aVar)
print("10 times 4:", 10 * 4)
anotherVar

4
10 times 4: 40


32

### Let's try!

You can run the code cell below to print the result.

In [15]:
numberStudents = 44
numberRooms = 4

# Calculate an even number of students per classroom
numberOfStudentsPerRoom = numberStudents / numberRooms

print(numberOfStudentsPerRoom)

11.0


Your turn!

Can you replace the ??? below and solve the following code?

In [16]:
pi = 3.14
radius = 12

# Can you calculate the perimeter of a circle using the two variables above?
# (Gentle reminder: P=2πr)
perimeter = 2 * pi * radius

print(perimeter)

75.36


After modifying the code above and running it, the output should be **75.36**

If you got it right, you can continue to the next section, *Types*.

## Basic Types

Not all data are equal!

In real life, you can't manipulate numbers the same way you manipulate words. You don’t use the same rules or syntax when you solve an equation or write a sentence.

That's exactly the same with programming languages!

That’s why Python uses **Types** to differentiate variables, so developers can manipulate different types of data.

Let's see the different **Types** existing in Python.

### Numbers

As we have seen in the previous section, Python supports two types of numbers - **integers** and **floating point** numbers.

**Integers** can be positive or negative, but don't have decimal values. **Floats** are the data type you use for manipulating decimal numbers.

In [17]:
myNegInt = -3
myPosInt = 7
myNegFloat = -1.37
myPosFloat = 9.99999

print(myNegInt)
print(myPosInt)
print(myNegFloat)
print(myPosFloat)

-3
7
-1.37
9.99999


Easy!

#### Operators

Integers and Floats support all the following arithmetic operations:

In [18]:
# Addition
print(10 + 3)  # 13

# Substraction
print(-3 - 10)  # -13

# Multiplication
print(6 * 7 * 2)  # 84

# Division (the result of a division is always a float)
print(10 / 3)  # 3.3333333
print(8 / 2)  # 4.0

# Modulo
# The modulo (%) operator returns the integer remainder of the division. dividend % divisor = remainder.
print(10 % 4)  # 2

# Power (double star)
print(2 ** 3)  # 8


13
-13
84
3.3333333333333335
4.0
2
8


In [19]:
# The order of operations follow the same rules as written equations.
# You can use parenthesis to group operations
# Can you tell what will be the result of the following equations before clicking on Run?
print(3 * 10 * 2 / 5)
print(10 / ((8 - 7) * 2))

12.0
5.0


### Booleans 

Booleans are specific types that reprensent a statement or a condition which can be **True** or **False**


In [20]:
myTrueBoolean = True
myFalseBoolean = False

print(myTrueBoolean)
print(myFalseBoolean)

# Dynamically set a boolean value from a condition
is2GreaterThan3 = 2 > 3
is2LowerThan3 = 2 < 3

print("is 2 greater than 3 ?", is2GreaterThan3)
print("is 2 lower than 3 ?", is2LowerThan3)

# Since a single = sign is used to assign value to a variable
# In order to check equality, we use a double == sign
# We can also group the condition inside parenthesis for better readability

is2Equal3 = (2 == 3)
is2Equal2 = (2 == 2)

print("is 2 equal 3 ?", is2Equal3)
print("is 2 equal 2 ?", is2Equal2)

# Inequality:  !=
is3DifferentThan100 = (3 != 100)
is3DifferentThan3 = (3 != 3)

print("is 3 different than 100 ?", is3DifferentThan100)
print("is 3 different than 3 ?", is3DifferentThan3)

# Greater or Equal / Lower or Equal:  >= / <=
is3GreaterEqualTo100 = (3 >= 100)
is3LowerEqualTo100 = (3 <= 100)
is3GreaterEqualTo3 = (3 >= 3)
is3LowerEqualTo3 = (3 <= 3)

print("is 3 greater or equal to 100 ?", is3GreaterEqualTo100)
print("is 3 lower or equal to 100 ?", is3LowerEqualTo100)
print("is 3 greater or equal to 3 ?", is3GreaterEqualTo3)
print("is 3 lower or equal to 3 ?", is3LowerEqualTo3)

# Inverse a boolean condition
print("not True is", not True)
print("not False is", not False)
print("not 2==2 is", not 2 == 2)

True
False
is 2 greater than 3 ? False
is 2 lower than 3 ? True
is 2 equal 3 ? False
is 2 equal 2 ? True
is 3 different than 100 ? True
is 3 different than 3 ? False
is 3 greater or equal to 100 ? False
is 3 lower or equal to 100 ? True
is 3 greater or equal to 3 ? True
is 3 lower or equal to 3 ? True
not True is False
not False is True
not 2==2 is False


#### Operators

OK, it gets a bit tricky here.

You can combine multiple conditions to make a full statement **True** or **False**, using two operators: **and** and **or**.

##### And

* All the conditions must be True for the whole statement to be True
* If at least one statement is False, then the whole statement is False

E.g.

In [21]:
print("(True and 2>3) returns", True and 2 > 3)  # False
print("(True and 2<3 and 2==2 and not False) returns", True and 2 < 3 and 2 == 2 and not False)  # True

(True and 2>3) returns False
(True and 2<3 and 2==2 and not False) returns True


##### Or

* If at least one condition is True, then the whole statement is True
* All the condition must be False for the whole statement to be False

E.g.

In [22]:
print("(False or 2==100 or False or 1==1) returns", (False) or (2 == 100) or (False) or (1 == 1))  # True
print("(not True or 2>3 or 2!=2 or False) returns", (not True) or (2 > 3) or (2 != 2) or (False))  # False

(False or 2==100 or False or 1==1) returns True
(not True or 2>3 or 2!=2 or False) returns False


### Strings

Strings represent words, sentences, names, everything that you can label.

They are declared using double quotes:

In [23]:
myString = "Hello!"
allCharactersAllowed = "Hi, my name is !@£$%^&*() and I'm 46 years old."
test = "ahsdasd \"asodiasd"

print(myString)
print(allCharactersAllowed)

# If you want to use double quotes in your sentence, you need to escape it using a anti-slash character "\" :

quotesInMyQuotes = "And he replied: \"Luke, I'm your father\""

print(quotesInMyQuotes)

Hello!
Hi, my name is !@£$%^&*() and I'm 46 years old.
And he replied: "Luke, I'm your father"


#### Operators

##### +

You can merge two or more Strings together to create a new one. This operation is called **Concatenating**.

To concatenate two or more Strings together, we use the `+` operator, because it's like "adding" Strings together.

In [24]:
helloWorld = "Hello" + " " + "world,"
my = " my "
nameIs = "name is"
print(helloWorld + my + nameIs + " Emily!")

Hello world, my name is Emily!


##### *

You can also "multiply" Strings, using the `*` operator. Not very useful though...

In [25]:
lotsOfHellos = "hello" * 10
print(lotsOfHellos)

hellohellohellohellohellohellohellohellohellohello


##### len() and [ ]

More useful, Strings are considered like lists of characters (we will learn about Lists in another section).

Which means we can access a specific character using its position (also called index).

We can also get the "size" or "length" of a String, which represents the number of characters a String is made of.

In [26]:
helloWorld = "Hello world!"
# We use square brackets to indicate the position of the letter we want to get.
# The position of the first letter of a String is 0, not 1!
print(helloWorld[6]) # Returns 'w'

# len() is the method used to return the length of a String
print("helloWorld is made of", len(helloWorld), "characters")

# If we combine everything, we can get the last character of a String dynamically:
# If it's too head-scratching, you can ask your teachers for help!
print("The last character of helloWorld is", helloWorld[len(helloWorld)-1]) 

w
helloWorld is made of 12 characters
The last character of helloWorld is !


### Casting

Are `1`, `1.0`, and `"1"` the same?

`1` is an Integer, `1.0` is a Float and `"1"` is a String. But, from a human perspective, the three of them just represent the number 1.

Let's find out if Python considers them equal:

In [28]:
print(1 == "1")
print(1 == 1.0)
print(1.0 == "1")

False
True
False


As you can see by running the code cell above, Python allows you to compare numbers (1 == 1.0 is True), but you can't compare numbers and Strings together (1.0 == "1" is False)!

This means, when you want to compare data, you must make sure all the data is from the same type. To do so, you need to convert data from a Type to another. This is called **Casting**.

In order to cast a value or a variable to a specific Type, you wrap them around the following methods:

In [29]:
intFromFloat = int(1.0)
intFromString = int("1")
print("1.0 and \"1\" are now both Integers:", intFromFloat == intFromString) 

floatFromInt = float(1)
floatFromString = float("1")
print("1 and \"1\" are now both Floats:", floatFromInt == floatFromString) 

stringFromInt = str(1)
stringFromFloat = str(1.0)
print("1 and 1.0 are now both Strings. But are they equal?", stringFromInt == stringFromFloat) 

1.0 and "1" are now both Integers: True
1 and "1" are now both Floats: True
1 and 1.0 are now both Strings. But are they equal? False


Hey?! Why are the two Strings above not equal??

Let's have a look at these two Strings in detail:

In [30]:
print("1 casted to String is:", str(1))
print("1.0 casted to String is:", str(1.0))

1 casted to String is: 1
1.0 casted to String is: 1.0


Remember when you learnt that a String is a list of characters?

Well, here, 

- `"1"` is a list of one character which is `1`.
- `"1.0"` is a list of three characters which are `1`, `.`, and `0`.

Therefore, these two Strings are not equal!

##### You can even cast Booleans!

1 is considered as True and 0 is considered as False.

Let's have a deeper look:

In [31]:
intToBool = bool(1)
floatToBool = bool(1.0)

print(intToBool, "is", floatToBool)

intToBool = bool(0)
floatToBool = bool(0.0)

print(intToBool, "is", floatToBool)

# From Strings, any non-empty String will be considered True, no matter what's written.
emptyStrToBool = bool("")
spaceStrToBool = bool(" ")
num1StrToBool = bool("1")
num0StrToBool = bool("0")
trueStrToBool = bool("True")
falseStrToBool = bool("False")

print("Empty String is", emptyStrToBool)
print("The rest is", spaceStrToBool, "and", num1StrToBool, "and", num0StrToBool, "and", trueStrToBool, "and", falseStrToBool)


True is True
False is False
Empty String is False
The rest is True and True and True and True and True


## Recap

Congratulations! You've learnt:

* The origins of Python
* How to create your library to load your lessons
* The different file types: iPython notebook (*.ipynb*) and Python script (*.py*)
* 3 methods to run Python code 
* Variables 
* Basic Types

## Now let's save your file

Select the file you've changed from the left side column. Right-click on mouse and select "Download", to save it to your local folder.

## References

* The magic of Python has been summarized into this poem [Tim Peters - The Zen of Python](https://www.python.org/dev/peps/pep-0020/#id3)
* [Variables and Types](https://www.learnpython.org/en/Variables_and_Types)

## Homework

Because what's a lesson without homework??????????????????????????????

Let's imagine you're feeling very thankful to Kelvin and you want to buy him breakfast. Here are some of his favourite breakfast items:

* Black sesame & cheese crescent (bread shop) - $2.20

* 2 soft-boiled eggs (kopitiam) - $1.00

* Plain prata (kopitiam) - $1.00

* Egg prata (kopitiam) - $1.50

* Kopi O kosong (kopitiam) - $0.90

* Tau huay with peanut rice balls (Mr. Bean) - $2.30

* Soy milk with 25% sugar (Mr. Bean) - $1.50

Q1: Create variables for all the above items relating to their cost

In [1]:
bread = 2.2
egg = 1
plain_prata = 1
egg_prata = 1.5
kopi = 0.9
tau_huay = 2.3
soy_milk = 1.5

Q2: How much would 2 plain pratas & 1 kopi o kosong cost? Pls factor in staff discount

Save the overall cost as a new variable called "not_enough" and print it out with some descriptive text (ie. "The cost of the breakfast will be $x.xx")

In [3]:
not_enough = (2+kopi)*0.8 #2 plain pratas @ the prata shop is actually sold at a discounted price of $1.80

print("The cost of the breakfast will be $", not_enough)

The cost of the breakfast will be $ 2.32


Q3: How much would a soy milk & 2 black sesame & cheese crescent cost? Pls factor in staff discount

Save the overall cost as a new variable called "still_not_enough" and print it out with some descriptive text (ie. "The cost of the breakfast will be $x.xx")

In [3]:
still_not_enough = 2*bread*0.8 + 1.5*0.9

print("The cost of the breakfast will be $", still_not_enough)

The cost of the breakfast will be $ 4.870000000000001


Q4: Imagine he is very hungry. How much would 2 black sesame & cheese crescent, 1 tau huay with peanut rice ball, 3 soy milks, and 10 egg pratas cost? Pls factor in staff discount

Save the overall cost as a new variable called "about_right" and print it out with some descriptive text (ie. "The cost of the breakfast will be $x.xx")

In [2]:
about_right = (2*bread + 10*egg_prata)*0.8 + (tau_huay + 3*soy_milk)*0.9

print("The cost of the breakfast will be $", about_right)

The cost of the breakfast will be $ 21.64
